client/src/main/java/io/pravega/client/stream/impl/ReaderGroupImpl.java

```diff
@@ -161,18 +164,43 @@ private Checkpoint completeCheckpoint(String checkpointName, StateSynchronizer<R
 
     @Override
     public void resetReadersToCheckpoint(Checkpoint checkpoint) {
+        resetReadersToStreamCut(new HashSet<>(checkpoint.asImpl().getPositions().values()));
+    }
+
+    @Override
+    public void resetReadersToStreamCut(final Set<StreamCut> streamCuts) {
         @Cleanup
         StateSynchronizer<ReaderGroupState> synchronizer = createSynchronizer();
+        synchronizer.fetchUpdates();
+
+        //ensure that streamCut for all the streams managed by the ReaderGroup are present.
+        Preconditions.checkArgument(validateStreamCuts(streamCuts, synchronizer.getState().getStreamNames()),
+                "StreamCuts for streams managed by the ReaderGroup need to be provided");
+
+        //reset the Readers to the StreamCut.
         synchronizer.updateState(state -> {
             ReaderGroupConfig config = state.getConfig();
             Map<Segment, Long> positions = new HashMap<>();
-            for (StreamCut cut : checkpoint.asImpl().getPositions().values()) {
+            for (StreamCut cut : streamCuts) {
                 positions.putAll(cut.getPositions());
             }
             return Collections.singletonList(new ReaderGroupStateInit(config, positions));
         });
     }
 
+    private boolean validateStreamCuts(final Collection<StreamCut> streamCuts, final Set<String> readerGroupStreams) {
+        //check if all the StreamCuts are unique and has no duplicates.
+        boolean isUnique = streamCuts.stream().map(sc -> sc.getStream().getStreamName()).allMatch(new HashSet<>()::add);
```

Comment:
This can be combined with the check below, which also creates a hashSet, by simply checking that readerGroupStreams.size() is the same as providedStreams.size()

Reply:
refactored it.